### ISSUE TITLE
fix(types): extend URL forbidden code point detection (C1, noncharacters) and fix vtab trim bug

### ISSUE THREAD

**Issue by YusukeHirao** [markuplint/markuplint#3629]
## Summary

Detect forbidden Unicode code points in HTML documents (primarily in attribute values).

## What to detect

| Code point range | Classification | Count in benchmark |
|-----------------|----------------|-------------------|
| U+FDD0–U+FDEF | Unicode noncharacters | 26 files |
| U+0080–U+009F | C1 control characters | 26 files |
| U+000B | Vertical Tab (C0) | 1 file |

Full forbidden set per spec:
- U+0001–U+0008, U+000E–U+001F (C0 controls, excluding TAB/LF/FF/CR)
- U+000B (Vertical Tab)
- U+007F–U+009F (DEL + C1 controls)
- U+FDD0–U+FDEF (noncharacters)
- U+FFFE, U+FFFF, U+1FFFE, U+1FFFF, ... (noncharacters at end of each plane)

## Context

All 53 test cases involve forbidden code points inside URL attribute values (`href`, `src`, `cite`, etc.), either in the host or path component.

## Spec reference

- https://html.spec.whatwg.org/multipage/parsing.html#preprocessing-the-input-stream

## Impact

~53 missed errors in the nu-html-checker compatibility benchmark.

**Comment by YusukeHirao** [comment]
### Design decision

Forbidden code points are **parse errors** per HTML LS, not attribute-level validation. The correct implementation is at the parser level.

**Approach**: Enable parse5's `onParseError` callback in `@markuplint/html-parser` and convert relevant parse errors into markuplint violations. parse5 already detects these code points but the handler is not configured.

**Required changes**:
1. `packages/@markuplint/html-parser/src/parser.ts` — Add `onParseError` to parse5 options
2. `packages/@markuplint/parser-utils/src/parser.ts` — Propagate parse errors through the AST
3. `packages/@markuplint/ml-core` — Convert parse errors to lint violations

This is a larger architectural change. Deferring to a future PR.

**Comment by YusukeHirao** [comment]
## nu-html-checker 実出力での検証

Docker 経由で nu-html-checker を実行して確認:

### テスト HTML
```html
<a href="http://example.com/[U+0085]">x</a>
<a href="http://example.com/[U+FDD0]">y</a>
<a href="http://example.com/[U+000B]">z</a>
```

### nu-html-checker 出力
6 件の error:
- 各 URL に対して "Forbidden code point" + "Bad value ... Illegal character in path segment"

### markuplint 出力
1 件の error のみ:
- `invalid-attr` で U+000B を検出
- **U+0085 (C1), U+FDD0 (noncharacter) は検出漏れ**

### 判定: CONFIRMED

C1 control range (U+0080–U+009F) と Unicode noncharacters (U+FDD0–U+FDEF) は未検出。`@markuplint/types` の URL validator 拡張が必要。

**Comment by YusukeHirao** [comment]
## ⚠️ 過去コメントは未検証

本 Issue の 2026-04-14 以前のコメントは nu-html-checker を実行せずに書かれた推測ベースです。以下が Docker 実測結果です。

## 実測結果（2026-04-15）

### 全 8 ケース対比（URL 属性値内の code point）

| # | Code point | 説明 | nu-html-checker | markuplint | Gap |
|---|-----------|------|----------------|-----------|-----|
| 1 | U+0080 | C1 先頭 | `Forbidden code point U+0080.` + `Illegal character in path segment` | （検出なし） | **YES** |
| 2 | U+009F | C1 末尾 | `Forbidden code point U+009f.` + Illegal character | （検出なし） | **YES** |
| 3 | U+000B | vtab (C0) | `Forbidden code point U+000b.` | （検出なし） | **YES**（リグレッション疑い） |
| 4 | U+FDD0 | noncharacter | `Forbidden code point U+fdd0.` + Illegal character | （検出なし） | **YES** |
| 5 | U+FDEF | noncharacter | `Forbidden code point U+fdef.` + Illegal character | （検出なし） | **YES** |
| 6 | U+FFFE | noncharacter | `Forbidden code point U+fffe.` + Illegal character | （検出なし） | **YES** |
| 7 | U+FFFF | noncharacter | `Forbidden code point U+ffff.` + Illegal character | （検出なし） | **YES** |
| 8 | U+00A0 | NBSP（有効） | 違反なし | 違反なし | NO |

## 重要な発見: リグレッション疑い

過去コメントで「U+000B は `invalid-attr` の URL validator で検出される」と記載されていたが、今回の実測では U+000B も**未検出**。環境差かリグレッションかは要追加確認。

## nu-html-checker の検出範囲

- C0 controls except TAB/LF/FF/CR: U+0001–U+0008, U+000B, U+000E–U+001F
- DEL + C1: U+007F–U+009F
- Plane noncharacters: U+FDD0–U+FDEF
- End-of-plane noncharacters: U+FFFE, U+FFFF, U+1FFFE, U+1FFFF, ..., U+10FFFE, U+10FFFF

## 実装方針

- `@markuplint/types` URL validator に code point 範囲チェック追加、または
- parse5 の `onParseError` 経由で parse 時に検出

## 真のスコープ

7/8 ケースが完全漏れ。Issue タイトル「forbidden Unicode code points in attribute values」は正確。

**Comment by YusukeHirao** [comment]
## 訂正: リグレッションではなく「部分実装 + trim バグ」

前コメントで「リグレッション疑い」と書いたが、コード調査と直接実測で真相が判明:

### 現状（2026-04-15 実測）

`@markuplint/types/src/whatwg/check-url.ts` に C0 control 検出コードは既に存在する:

```typescript
const C0_CONTROL = /[\u0000-\u0008\v\u000E-\u001F\u007F]/;
```

ただし 2 つの欠陥がある:

#### 欠陥 1: regex の対象範囲不足

現在の `C0_CONTROL` は C0 controls と DEL のみカバー。以下は**未実装**:

- **C1 controls** U+0080–U+009F
- **Plane noncharacters** U+FDD0–U+FDEF
- **End-of-plane noncharacters** U+FFFE, U+FFFF, ..., U+10FFFE, U+10FFFF

nu-html-checker はこれらも `Forbidden code point U+XXXX` として報告する。

#### 欠陥 2: `.trim()` によるサイレント除去バグ

check-url.ts の `value.trim()` が URL 末尾の U+000B (vtab) を削除してしまう。JavaScript の `String.prototype.trim()` は U+000B を whitespace として扱うため、C0 チェックに到達する前に消える。

##### 実証

| 入力 | 結果 |
|------|------|
| `<a href="http://example.com/\v">` (末尾) | **検出されない**（trim で削除） |
| `<a href="http://exa\vmple.com/">` (途中) | 検出される（「期待されない文字を含んでいます」） |

### 最初の Agent 誤報の原因

最初の Agent は URL 途中に U+000B を置くテストをしていた可能性があり、その場合は検出される。詳細 Agent のテストは末尾配置で、trim に削除されるため未検出。どちらも実装の挙動どおりで、リグレッションではない。

### 正しいスコープ

#3629 の実装対象は 2 つ:

1. **regex 拡張**: `C0_CONTROL` を C1 + noncharacters を含む `FORBIDDEN_CODE_POINT` に拡張（または別 regex 追加）
2. **trim バグ修正**: URL 末尾の vtab が trim で消えないよう前処理ロジックを修正（`trim()` ではなく specific whitespace set で切り出す等）
